packages/utils/src/abi_utils.ts

```diff
@@ -1,7 +1,166 @@
+
+export type EthersParamName = null | string | EthersNestedParamName;
+
+export interface EthersNestedParamName {
+    name: string | null;
+    names: EthersParamName[];
+}
+
+// Note(albrow): This function is unexported in ethers.js. Copying it here for
+// now.
+// Source: https://github.com/ethers-io/ethers.js/blob/884593ab76004a808bf8097e9753fb5f8dcc3067/contracts/interface.js#L30
+function parseEthersParams(params: DataItem[]): { names: EthersParamName[]; types: string[] } {
+    const names: EthersParamName[] = [];
+    const types: string[] = [];
+
+    params.forEach((param: DataItem) => {
+        if (param.components != null) {
+            let suffix = '';
+            const arrayBracket = param.type.indexOf('[');
+            if (arrayBracket >= 0) {
+                suffix = param.type.substring(arrayBracket);
+            }
+
+            const result = parseEthersParams(param.components);
+            names.push({ name: param.name || null, names: result.names });
+            types.push('tuple(' + result.types.join(',') + ')' + suffix);
+        } else {
+            names.push(param.name || null);
+            types.push(param.type);
+        }
+    });
+
+    return {
+        names,
+        types,
+    };
+}
+
+// returns true if x is equal to y and false otherwise. Performs some minimal
+// type conversion and data massaging for x and y, depending on type. name and
+// type should typically be derived from parseEthersParams.
+function isAbiDataEqual(name: EthersParamName, type: string, x: any, y: any): boolean {
+    if (_.isUndefined(x) && _.isUndefined(y)) {
+        return true;
+    } else if (_.isUndefined(x) && !_.isUndefined(y)) {
+        return false;
+    } else if (!_.isUndefined(x) && _.isUndefined(y)) {
+        return false;
+    }
+    if (_.endsWith(type, '[]')) {
+        // For array types, we iterate through the elements and check each one
+        // individually. Strangely, name does not need to be changed in this
+        // case.
+        if (x.length !== y.length) {
+            return false;
+        }
+        const newType = _.trimEnd(type, '[]');
+        for (let i = 0; i < x.length; i++) {
+            if (!isAbiDataEqual(name, newType, x[i], y[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+    if (_.startsWith(type, 'tuple(')) {
+        if (_.isString(name)) {
+            throw new Error('Internal error: type was tuple but names was a string');
+        } else if (_.isNull(name)) {
+            throw new Error('Internal error: type was tuple but names was a null');
+        }
+        // For tuples, we iterate through the underlying values and check each
+        // one individually.
+        const types = splitTupleTypes(type);
+        if (types.length !== name.names.length) {
+            throw new Error(
+                `Internal error: parameter types/names length mismatch (${types.length} != ${name.names.length})`,
```

Comment:
What other error prefixes should we use? And when exactly is coming considered an `Internal error`?

Reply:
This is what I was asking when I mentioned error-handling conventions. To me, an "internal error" is due to a bug in the implementation itself as opposed to invalid input or network/database errors.

I think we should standardize the kinds of error messages we return in our libraries, possibly by including prefixes, but that lies outside of the scope of this PR. Is it okay to leave this for now and we can circle back on it in the future? Or should I change this error message right now to make it more clear?

Reply:
Yep, that's fine. I'm down to standardize them though, would be great if you could take the lead on that.